scm_v3c/applications/sensor_adc/svd.c

```diff
@@ -0,0 +1,120 @@
+
+// ONLY FOR 2X2 MATRICES
+
+#include "svd.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "matrix.h"
+
+matrix_t mtx;
+
+bool svd_init(matrix_t* m) {
+    if (m->cols != 2 || m->rows != 2) {
+        return false;
+    }
+    matrix_init(&mtx, 2, 2);
+    matrix_copy(m, &mtx);
+    return true;
+}
+
+// calculates the square root of s with tolerance 0.0001
+// https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Heron's_method 
+matrix_type_t herons_method(matrix_type_t s) {
+    matrix_type_t guess = s;
+    while ((guess * guess) - s > 0.0001 * s) {
+        guess = (guess + (s / guess)) / 2;
+    }
+    return guess;
+}
+
+// calculates the square of a number n
+matrix_type_t square(matrix_type_t n) {
+    if (n < 0) {
+        matrix_type_t positive_n = -n;
+        return positive_n * positive_n;
+    } else {
+        return n * n;
+    }
+}
+
+// calculate the eigenvalues of a 2x2 matrix with entries a, b, c, and d
+void eigenvalues(matrix_type_t a, matrix_type_t b, 
+                matrix_type_t c, matrix_type_t d,
+                matrix_type_t* eigenvalue_1, matrix_type_t* eigenvalue_2) {
+    *eigenvalue_1 = (a + d + herons_method(square(a + d) - (4 * (square(a) - square(b))))) / 2;
+    *eigenvalue_2 = (a + d - herons_method(square(a + d) - (4 * (square(a) - square(b))))) / 2;
+}
+
+// calculate the nullspace of a 2x2 matrix
+// https://www.quora.com/How-do-you-find-the-null-space-of-a-2x2-matrix 
+void nullspace(matrix_type_t a, matrix_type_t b, 
```

Comment:
how are you handling the all-zero matrix with dim of nullspace = 2?  are you only returning one nullspace vector?